### PR TITLE
Improve topology python scripts for ipc4

### PR DESCRIPTION
**Pull request by aiChaoSONG**
Yongan has a new version of tplgtool2.py, and we are supposed to do a new sof-tplgreader.py based on tplgtool2.py, however, no one has time for it. the old sof-tplgreader.py and tplgtool.py are still used.
So this patch does the improvement based on old python scripts, to make the pipeline filter better for IPC4. 

The main purpose is to filter out wov capture pipeline in normal capturing. so this is needed before adding wov pipeline to ipc4 nocodec topology

### CODE REVIEW

**Review by paulstelian97 — APPROVED** [review]

**Comment by marc-hb** [comment]
I believe `tplgtool2.py` is already used in CI every day to generate graphviz output, correct?

> however, no one has time for it. the old sof-tplgreader.py and tplttool.py are still used.
So this patch does the improvement based on old python scripts, to make the pipeline filter better for IPC4.

This PR looks small enough.

**Comment by marc-hb** [comment]
The errors in https://sof-ci.01.org/softestpr/PR1008/build168/devicetest/index.html looks relevant (and cryptic)

**Comment by aiChaoSONG** [comment]
> The errors in https://sof-ci.01.org/softestpr/PR1008/build168/devicetest/index.html look relevant (and cryptic)

  It could be caused by the complex smart_amp topology, looks there is issue to find host widgets for smart amp playback and capture.

File: tools/sof-tplgreader.py [hunk 1]
```diff
@@ -361,14 +363,15 @@ def parse_and_set_block_keyword(block_str, tplgreader):
...
-    default_block_keyword = 'kpbm:any;type:capture & echo;'
+    default_block_keyword = 'kpbm:any;kpb:any;type:capture & echo;'
     # if no filter or block item is specified, the "block_none" here will help to
     # block nothing
     block_str = "block_none"
     # user specified block items from command line
     cmd_block_str = ret_args['block'].strip() if ret_args['block'] is not None else ''
 
-    if ret_args['filter'] is not None and 'kpbm' not in ret_args['filter'] and 'echo' not in ret_args['filter']:
+    if ret_args['filter'] is not None and 'echo' not in ret_args['filter'] and \
```
Comment (marc-hb) [inline tools/sof-tplgreader.py:363]: Are you sure you need  the backslash?

This is how I typically handle situations like this:

- apt install black
- cp sof-tplgreader.py for_black.py
- black for_black.py
- copy line 373 and 374 from for_black.py to sof-tplgreader.py

**Review by marc-hb — COMMENTED** [review]
The Python code looks "clean" but I don't know enough about topologies to review it.

Given the choice reverts and other fixes usually go first. Bug fixes before new features.

File: tools/sof-tplgreader.py [hunk 2]
```diff
@@ -361,14 +353,19 @@ def parse_and_set_block_keyword(block_str, tplgreader):
...
     block_str = "block_none"
     # user specified block items from command line
     cmd_block_str = ret_args['block'].strip() if ret_args['block'] is not None else ''
 
-    if ret_args['filter'] is not None and 'kpbm' not in ret_args['filter'] and 'echo' not in ret_args['filter']:
+    if (
+        ret_args["filter"] is not None
+        and "echo" not in ret_args["filter"]
+        and "kpbm" not in ret_args["filter"]
+        and "kpb" not in ret_args["filter"]
```
Comment (marc-hb) [inline tools/sof-tplgreader.py:367]:
```suggestion
                        and not set.intersection({"echo", "kpbm", "kpb"}, ret_args["filter"])
```
Even better: give the `{"echo", "kpbm", "kpb"}` `set` a name, something like `[in]compatible_features`

**Comment by marc-hb** [comment]
Irrelevant suspend/resume failures in https://sof-ci.01.org/softestpr/PR1008/build187/devicetest/index.html

Other failures in https://sof-ci.01.org/softestpr/PR1008/build188/devicetest/index.html

File: tools/sof-tplgreader.py (hunk 2 above)
Comment (aiChaoSONG) [reply tools/sof-tplgreader.py:367]: ret_args["filter"] is a string from command line, the value for -f option below.
`sof-tplgreader.py /usr/lib/firmware/intel/sof-tplg/sof-hda-generic-4ch.tplg -f 'type:playback & id:any' -b ' pcm:HDA Digital' -s 0 -e`, so set is not going to work here. How about a re match with pattern 'echo|kpbm|kpb'.

**Comment by aiChaoSONG** [comment]
SOFCI TEST

**Comment by aiChaoSONG** [comment]
@marc-hb  Ping

**Comment by marc-hb** [comment]
@aiChaoSONG :

> The Python code looks "clean" but I don't know enough about topologies to review it.

**Review by keqiaozhang — APPROVED** [review]
LGTM

**Review by miRoox — APPROVED** [review]
LGTM

**Pull request merged**